File: Eudoxus.Portal/Controls/BotShield.js
function ChangeBotShieldInput(Obj) {
    var change = 0;
    var Str = Obj.value;
    var StrL = Str.length;
    var StrOut = '';

    var c = ' ';
    var cOut = ' ';

    for (var i = 0; i < StrL; i++) {
        c = Str.substring(i, i + 1);
        cOut = '';

        if ((c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9')) { cOut = c; }
        if (c == 'Α' || c == 'a' || c == 'α' || c == 'ά') { change = 1; cOut = 'A'; }
        if (c == 'Β' || c == 'b' || c == 'β') { change = 1; cOut = 'B'; }
        if (c == 'Γ' || c == 'g' || c == 'γ') { change = 1; cOut = 'G'; }
        if (c == 'Δ' || c == 'd' || c == 'δ') { change = 1; cOut = 'D'; }
        if (c == 'Ε' || c == 'e' || c == 'ε' || c == 'έ') { change = 1; cOut = 'E'; }
        if (c == 'Ζ' || c == 'z' || c == 'ζ') { change = 1; cOut = 'Z'; }
        if (c == 'Η' || c == 'h' || c == 'η' || c == 'ή') { change = 1; cOut = 'H'; }
        if (c == 'Θ' || c == 'u' || c == 'θ') { change = 1; cOut = 'U'; }
        if (c == 'Ι' || c == 'Ϊ' || c == 'i' || c == 'ι' || c == 'ί' || c == 'ϊ' || c == 'ΐ') { change = 1; cOut = 'I'; }
        if (c == 'Κ' || c == 'k' || c == 'κ') { change = 1; cOut = 'K'; }
        if (c == 'Λ' || c == 'l' || c == 'λ') { change = 1; cOut = 'L'; }
        if (c == 'Μ' || c == 'm' || c == 'μ') { change = 1; cOut = 'M'; }
        if (c == 'Ν' || c == 'n' || c == 'ν') { change = 1; cOut = 'N'; }
        if (c == 'Ξ' || c == 'j' || c == 'ξ') { change = 1; cOut = 'J'; }
        if (c == 'Ο' || c == 'o' || c == 'ο' || c == 'ό') { change = 1; cOut = 'O'; }
        if (c == 'Π' || c == 'p' || c == 'π') { change = 1; cOut = 'P'; }
        if (c == 'Ρ' || c == 'r' || c == 'ρ') { change = 1; cOut = 'R'; }
        if (c == 'Σ' || c == 's' || c == 'σ') { change = 1; cOut = 'S'; }
        if (c == 'Τ' || c == 't' || c == 'τ') { change = 1; cOut = 'T'; }
        if (c == 'Υ' || c == 'Ϋ' || c == 'y' || c == 'υ' || c == 'ύ' || c == 'ϋ' || c == 'ΰ') { change = 1; cOut = 'Y'; }
        if (c == 'Φ' || c == 'f' || c == 'φ') { change = 1; cOut = 'F'; }
        if (c == 'Χ' || c == 'x' || c == 'χ') { change = 1; cOut = 'X'; }
        if (c == 'Ψ' || c == 'c' || c == 'ψ') { change = 1; cOut = 'C'; }
        if (c == 'Ω' || c == 'v' || c == 'ω' || c == 'ώ') { change = 1; cOut = 'V'; }
        if (c == 'Q' || c == 'q') { change = 1; cOut = 'Q'; }
        if (c == 'w' || c == 'ς') { change = 1; cOut = 'W'; }
        if (c == ';') { change = 1; cOut = 'Q'; }
        if (cOut == '') { change = 1; }

        StrOut = StrOut + cOut;
    };

    if (change == 1) Obj.value = StrOut;
    return;
}

function ValidateBotShieldInput(s, e) {
    e.IsValid = e.Value != '';
    s.errormessage = DEFAULT_BS_ERROR_MESSAGE;
    if (e.IsValid) {
        if (e.Value.length != BS_MAX_LENGTH) {
            e.IsValid = false;
            s.errormessage = 'Το μήκος των χαρακτήρων πρέπει να είναι ακριβώς ' + BS_MAX_LENGTH;
        }
    }
}

File: Eudoxus.Portal/Controls/BotShield.cs
using System;
using System.Web.UI;
using System.Web.UI.WebControls;
using Lanap.BotDetect;

namespace Eudoxus.Portal.Controls
{
    [ToolboxData("<{0}:BotShield runat=server></{0}:BotShield>")]
    public class BotShield : WebControl, INamingContainer
    {
        private ITemplate contentTemplate;

        [TemplateContainer(typeof(BotShield))]
        [PersistenceMode(PersistenceMode.InnerProperty)]
        public ITemplate ContentTemplate
        {
            get { return contentTemplate; }
            set { contentTemplate = value; }
        }

        protected virtual ITemplate CreateDefaultTemplate()
        {
            return new LanapTemplate();
        }

        protected override void CreateChildControls()
        {
            Controls.Clear();
            ITemplate tmpl = contentTemplate != null ? contentTemplate : CreateDefaultTemplate();
            tmpl.InstantiateIn(this);
        }

        protected override HtmlTextWriterTag TagKey
        {
            get
            {
                return HtmlTextWriterTag.Div;
            }
        }

        protected override void OnPreRender(EventArgs e)
        {
            base.OnPreRender(e);
            Captcha c = FindControl(SHIELD_ID) as Captcha;
            if (c != null)
            {
                Array a = Enum.GetValues(typeof(TextStyleEnum));
                Random r = new Random();
                int x = r.Next(a.Length);
                c.TextStyle = (TextStyleEnum)a.GetValue(x);
            }
            if (this.Page != null && !this.Page.ClientScript.IsClientScriptIncludeRegistered(this.GetType(), SCRIPT_KEY))
            {
                this.Page.ClientScript.RegisterClientScriptInclude(this.GetType(), SCRIPT_KEY, this.Page.ClientScript.GetWebResourceUrl(this.GetType(), "Eudoxus.Portal.Controls.BotShield.js"));

                if (c != null)
                    this.Page.ClientScript.RegisterClientScriptBlock(this.GetType(), SCRIPT_KEY + "1", string.Format("var BS_MAX_LENGTH = {0}; var DEFAULT_BS_ERROR_MESSAGE = '{1}';", c.CodeLength, DEFAULT_ERROR_MESSAGE), true);
            }


            if (Page.IsPostBack)
            {
                BaseValidator val = FindControl(VALIDATOR_ID) as BaseValidator;
                if (!val.IsValid)
                {
                    //ForceDraw();
                    ITextControl txt = FindControl(TEXT_ID) as ITextControl;
                    txt.Text = "";
                }
            }
        }

        public string ValidationGroup
        {
            get
            {
                return (FindControl(VALIDATOR_ID) as BaseValidator).ValidationGroup;
            }
            set { (FindControl(VALIDATOR_ID) as BaseValidator).ValidationGroup = value; }
        }

        public const string TEXT_ID = "txtFormShield";
        public const string SHIELD_ID = "formShield";
        public const string VALIDATOR_ID = "valFormShield";

        public const string SCRIPT_FUNCTION = "ChangeBotShieldInput";
        public const string DEFAULT_ERROR_MESSAGE = "<b>Παρακαλώ συμπληρώστε το κείμενο που εμφανίζεται στην εικόνα</b>";
        public const string SCRIPT_VALIDATION_FUNCTION = "ValidateBotShieldInput";
        public const string SCRIPT_KEY = "Eudoxus.Portal.Controls.ChangeBotShieldInput";
    }

    sealed class LanapTemplate : ITemplate
    {
        private Literal lt = new Literal();
        private Table tbl = new Table();
        private Captcha captcha = new Captcha();
        private TextBox txtFormShield = new TextBox();
        private CustomValidator valFormShield = new CustomValidator();

        internal LanapTemplate()
        {
            lt.Text = "<p><b>Γράψτε τους χαρακτήρες που εμφανίζονται στην εικόνα</b></p>";

            TableRow tr;
            TableCell tc;

            tbl.Rows.Add(tr = new TableRow());

            tr.Cells.Add(tc = new TableCell());
            captcha.ID = BotShield.SHIELD_ID;
            captcha.SoundEnabled = false;
            captcha.ReloadEnabled = true;
            captcha.ReloadIconAltText = "Αλλαγή εικόνας";
            tc.Controls.Add(captcha);

            tr.Cells.Add(tc = new TableCell());
            txtFormShield.ID = BotShield.TEXT_ID;
            txtFormShield.Attributes["onkeyup"] = BotShield.SCRIPT_FUNCTION + "(this)";
            tc.Controls.Add(txtFormShield);

            tbl.Rows.Add(tr = new TableRow());
            tr.Cells.Add(tc = new TableCell());
            tc.ColumnSpan = 2;
            valFormShield.ID = BotShield.VALIDATOR_ID;
            valFormShield.ErrorMessage = BotShield.DEFAULT_ERROR_MESSAGE;
            valFormShield.ClientValidationFunction = BotShield.SCRIPT_VALIDATION_FUNCTION;
            valFormShield.ControlToValidate = txtFormShield.ClientID;
            valFormShield.SetFocusOnError = false;
            valFormShield.ValidateEmptyText = true;
            valFormShield.ServerValidate += delegate(object sender, ServerValidateEventArgs e)
            {
                Control ctl = ((Control)sender).Parent;
                e.IsValid = ((Captcha)ctl.FindControl(BotShield.SHIELD_ID)).Validate(
                    ((TextBox)ctl.FindControl(BotShield.TEXT_ID)).Text);
            };

            tc.Controls.Add(valFormShield);
        }

        public void InstantiateIn(Control container)
        {
            container.Controls.Add(lt);
            container.Controls.Add(tbl);
        }
    }
}
